File: lib/easy/tree_path_has_sum.dart
import 'package:dark_coding_challenges/helpers/data_structures/tree_node.dart';

bool treePathHasSum(TreeNode? root, int targetSum) {
  if (root == null) return false;
  final value = root.val;
  if (value == null) return false;
  int remainder = targetSum - value;
  return treePathHasSum(root.left, remainder) || treePathHasSum(root.right, remainder);
}

File: src/easy/tree_path_has_sum.rs
use crate::helpers::tree_node::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

pub fn tree_path_has_sum(root: Option<Rc<RefCell<TreeNode<i32>>>>, target_sum: i32) -> bool {
    match root {
        None => false,
        Some(current) => {
            let node = current.borrow();
            let remainder = target_sum - node.val;
            if node.left.is_none() && node.right.is_none() {
                return target_sum == node.val;
            }
            tree_path_has_sum(node.left.clone(), remainder)
                || tree_path_has_sum(node.right.clone(), remainder)
        }
    }
}

